src/vs/workbench/contrib/notebook/browser/contrib/notebookVariables/notebookVariablesDataSource.ts

```diff
@@ -0,0 +1,94 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the MIT License. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+import { IAsyncDataSource } from 'vs/base/browser/ui/tree/tree';
+import { CancellationToken } from 'vs/base/common/cancellation';
+import { NotebookTextModel } from 'vs/workbench/contrib/notebook/common/model/notebookTextModel';
+import { INotebookKernelService, VariablesResult } from 'vs/workbench/contrib/notebook/common/notebookKernelService';
+
+export interface INotebookScope {
+	type: 'root';
+	readonly notebook: NotebookTextModel | undefined;
+}
+
+export interface INotebookVariableElement {
+	type: 'variable';
+	readonly id: number;
+	readonly name: string;
+	readonly value: string;
+	readonly indexedChildrenCount: number;
+	readonly hasNamedChildren: boolean;
+}
+
+export class NotebookVariableDataSource implements IAsyncDataSource<INotebookScope, INotebookVariableElement> {
+
+	private notebook: NotebookTextModel | undefined = undefined;
+
+	constructor(private readonly notebookKernelService: INotebookKernelService) { }
+
+	hasChildren(element: INotebookScope | INotebookVariableElement): boolean {
+		return element.type === 'root' || element.hasNamedChildren || element.indexedChildrenCount > 0;
+	}
+
+	async getChildren(element: INotebookScope | INotebookVariableElement): Promise<Array<INotebookVariableElement>> {
+		if (element.type === 'root') {
+			this.notebook = element.notebook;
```

Comment:
Instead of having this state on the data source object, it might be nicer if all the state needed to look up child variables is on the variable itself.